File: SecureBlackbox Samples/Office Verifier/python/office_verifier.py
# 
# SecureBlackbox 2022 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of SecureBlackbox in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/secureblackbox
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from secureblackbox import *

input = sys.hexversion<0x03000000 and raw_input or input

def ensureArg(args, prompt, index):
  if len(args) <= index:
    while len(args) <= index:
      args.append(None)
    args[index] = input(prompt)
  elif args[index] == None:
    args[index] = input(prompt)

def displayHelp():
    print(
        "NAME\n"
        "  officeverifier -- SecureBlackbox OfficeVerifier Demo Application\n\n"
        "SYNOPSIS\n"
        "  officeverifier [-input input_file] [-cert certificate_file] [-certpass certificate_password]\n"
        "DESCRIPTION\n"
        "  OfficeVerifier demonstrates the usage of OfficeVerifier from SecureBlackbox.\n"
        "  Used to verify the signature.\n\n"
        "  The options are as follows:\n\n"
        "  -input        An input file to verify (Required).\n\n"
        "  -cert         The certificate used to sign files.\n\n"
        "  -certpass     The password for the signing certificate.\n\n"
        "  -performRevocationCheck       Whether certificate revocation information should be checked.\n\n"
        "  -ignoreChainValidationErrors  Whether to ignore chain validation errors.\n\n"
        "  -forceCompleteChainValidation Whether to check issuer (CA) certificates when the signing certificate is invalid.\n\n"
    )
    
def translateSigType(value):
    if (value == 1): # ostBinaryCryptoAPI
        return "BinaryCryptoAPI"
    elif (value == 2): # ostBinaryXML
        return "BinaryXML"
    elif (value == 3): # ostOpenXML
        return "OpenXML"
    elif (value == 4): # ostOpenXPS
        return "OpenXPS"
    elif (value == 5): # ostOpenDocument
        return "OpenOffice"
    else:
        return "Unknown"

def translateDocSig(value):
    if (value):
        return "Document content is signed"
    else:
        return "Document content is partially signed"

def translateCore(value):
    if (value):
        return "Document properties are signed"
    else:
        return "Document properties are not signed"

def translateOrigSig(value):
    if (value):
        return "Signature origin is signed"
    else:
        return "Signature origin is not signed"

def translateSigValidity(value):
    if (value == svtValid):
        return "Valid"
    elif (value == svtCorrupted):
        return "Corrupted"
    elif (value == svtSignerNotFound):
        return "Signer not found"
    elif (value == svtFailure):
        return "Failure"
    else:
        return "Unknown" 
    
def translateChainValidity(value):
    if (value == cvtValid):
        return "Valid"
    elif (value == cvtValidButUntrusted):
        return "Valid but untrusted"
    elif (value == cvtInvalid):
        return "Invalid"
    elif (value == cvtCantBeEstablished):
        return "Can't be established"
    else:
        return "Unknown" 
    
if (len(sys.argv) < 3):
    displayHelp()
    sys.exit(1)
else:
    verifier = OfficeVerifier()
    cm = CertificateManager()
    
    inputF = ""
    certF = ""
    certpass = ""
    performRevocationCheck = False
    ignoreChainValidationErrors = False
    forceCompleteChainValidation = False
    
    try:
        for x in range(len(sys.argv)):
            if (sys.argv[x].startswith("-")):
                if (sys.argv[x].lower() == "-input"):
                    inputF = sys.argv[x+1]
                if (sys.argv[x].lower() == "-cert"):
                    certF = sys.argv[x+1]
                if (sys.argv[x].lower() == "-certpass"):
                    certpass = sys.argv[x+1]
                if (sys.argv[x].lower() == "-performRevocationCheck"):
                    performRevocationCheck = True
                if (sys.argv[x].lower() == "-ignoreChainValidationErrors"):
                    ignoreChainValidationErrors = True
                if (sys.argv[x].lower() == "-forceCompleteChainValidation"):
                    forceCompleteChainValidation = True
                    
        if (inputF == ""):
            print("-input is required.\n")
            displayHelp()
            sys.exit(1)

        verifier.set_input_file(inputF)
        
        
        if performRevocationCheck:
            verifier.set_revocation_check(1) # Auto
        else:
            verifier.set_revocation_check(0) # None
        
        if ignoreChainValidationErrors:
            verifier.set_ignore_chain_validation_errors(True) 
        else:
            verifier.set_ignore_chain_validation_errors(False)
            
        if forceCompleteChainValidation:
            verifier.config("ForceCompleteChainValidation=true")
        else:
            verifier.config("ForceCompleteChainValidation=false")       
        
        if not (certF == ""):
            cm.import_from_file(certF, certpass)
            verifier.set_known_cert_count(1)
            verifier.set_known_cert_handle(0, cm.get_cert_handle())
            
        verifier.verify()
        
        print("There are %i signatures in this file.\n"%(verifier.get_signature_count()))
        for x in range(verifier.get_signature_count()):
            print("Signature " + str(x+1))
            print("  Signature type: " + translateSigType(verifier.get_signature_signature_type(x)))
            print("  " + translateDocSig(verifier.get_signature_document_signed(x)))
            print("  " + translateCore(verifier.get_signature_core_properties_signed(x)))
            print("  " + translateOrigSig(verifier.get_signature_signature_origin_signed(x)))
            print("  Signature Time: " + verifier.get_signature_signature_time(x))
            print("  Signature Validation Result: " + translateSigValidity(verifier.get_signature_signature_validation_result(x)))
            print("  Chain Validation Result: " + translateChainValidity(verifier.get_signature_chain_validation_result(x)) + "\n")
    except Exception as e: 
        print(e)

File: SecureBlackbox Samples/Office Verifier/php/office_verifier.php
<?php
/*
 * SecureBlackbox 2024 PHP Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
require_once('../include/secureblackbox_officeverifier.php');
require_once('../include/secureblackbox_certificatemanager.php');
require_once('../include/secureblackbox_const.php');
?>
<style>
table { width: 100% !important; }
td { white-space: nowrap; }
td input { width: 100%; }
td:last-child { width: 100%; }
</style>

<div width="90%">
  <form method=POST>
    <h2>Office Verifier Demo</h2>
    
    <b>General Options</b><br/><br/>
    <table>
      <tr><td>Input File:</td><td><input type=text name=inputFile value=""></td></tr>
    </table>

    <br/><br/><b>Validation Options</b><br/><br/>

    <input type=checkbox name="ignoreChainValidationErrors" /><label for=ignoreChainValidationErrors>Ignore chain validation errors</label>
    <input type=checkbox name="forceCompleteChainValidation" /><label for=forceCompleteChainValidation>Force complete chain validation</label>
    <input type=checkbox name="performRevocationCheck" /><label for=performRevocationCheck>Perform revocation check</label>

    <br/><br/><br/>

    <b>Additional Certificates</b>
    <p>Enter the path(s) for any certificate(s), one per line. If a password is required add a semicolon followed by the password (e.g. C:\path\to\my.pfx;password).</p>
    <table>
      <tr><td>Known Certificates:</td><td><textarea style="font-family: Arial, sans-serif; width: 100%" name=knownCerts rows=10></textarea></td></tr>
      <tr><td>Trusted Certificates:</td><td><textarea style="font-family: Arial, sans-serif; width: 100%" name=trustedCerts rows=10></textarea></td></tr>
    </table>

    <input type="submit" value="Verify" />
  </form>
</div><br/>

<?php

function translateSignatureType($sigType){
  switch($sigType){
    case OFFICEVERIFIER_SIGNATURESIGNATURETYPE_BINARY_CRYPTO_API:  return "BinaryCryptoAPI";     break;
    case OFFICEVERIFIER_SIGNATURESIGNATURETYPE_BINARY_XML:         return "BinaryXML";     break;
    case OFFICEVERIFIER_SIGNATURESIGNATURETYPE_OPEN_XML:           return "OpenXML"; break;
    case OFFICEVERIFIER_SIGNATURESIGNATURETYPE_OPEN_XPS:           return "OpenXPS";     break;
    case OFFICEVERIFIER_SIGNATURESIGNATURETYPE_OPEN_DOCUMENT:      return "OpenOffice"; break;
    default: return "Unknown";   break;
  }
}

function translateDocSig($value) {
  if ($value) 
  {
    return "Document content is signed";
  }
  else
  {
    return "Document content is partially signed";
  }
}

function translateCore($value) {
  if ($value)
  {
    return "Document properties are signed";
  }
  else
  {
    return "Document properties are not signed";
  }
}

function translateOrigSig($value) {
  if ($value)
  {
    return "Signature origin is signed";
  }
  else
  {
    return "Signature origin is not signed";
  }
}

function translateSigValidationResult($value){
  switch($value){
    case OFFICEVERIFIER_SIGNATURESIGNATUREVALIDATIONRESULT_VALID:             return "Valid";     break;
    case OFFICEVERIFIER_SIGNATURESIGNATUREVALIDATIONRESULT_CORRUPTED:         return "Corrupted";     break;
    case OFFICEVERIFIER_SIGNATURESIGNATUREVALIDATIONRESULT_SIGNER_NOT_FOUND:  return "Signer not found"; break;
    case OFFICEVERIFIER_SIGNATURESIGNATUREVALIDATIONRESULT_FAILURE:           return "Failure";     break;
    default: return "Unknown";   break;
  }
}
function translateChainValidationResult($value){
  switch($value){
    case OFFICEVERIFIER_SIGNATURECHAINVALIDATIONRESULT_VALID:                return "Valid";     break;
    case OFFICEVERIFIER_SIGNATURECHAINVALIDATIONRESULT_VALID_BUT_UNTRUSTED:  return "Valid but untrusted";     break;
    case OFFICEVERIFIER_SIGNATURECHAINVALIDATIONRESULT_INVALID:              return "Invalid"; break;
    case OFFICEVERIFIER_SIGNATURECHAINVALIDATIONRESULT_CANT_BE_ESTABLISHED:  return "Can't be established";     break;
    default: return "Unknown";   break;
  }
}

  if ($_SERVER['REQUEST_METHOD'] == "POST") {
    $officeverifier = new SecureBlackbox_OfficeVerifier();
    $certmgr = new SecureBlackBox_CertificateManager();

    try {
      // General options
      $officeverifier->setInputFile($_REQUEST['inputFile']);

      // Additional options
      $officeverifier->setIgnoreChainValidationErrors(!empty($_REQUEST['ignoreChainValidationErrors']));
      $officeverifier->doConfig("ForceCompleteChainValidation=" . (!empty($_REQUEST['forceCompleteChainValidation']) ? "true" : "false"));
      $officeverifier->setRevocationCheck(!empty($_REQUEST['performRevocationCheck']) ? OFFICEVERIFIER_REVOCATIONCHECK_AUTO : OFFICEVERIFIER_REVOCATIONCHECK_NONE);

      // Known certificates
      $certPaths = trim($_REQUEST['knownCerts']);
      if (strlen($certPaths) > 0) {
        $knownCerts = explode("\r\n", $certPaths);
        $officeverifier->setKnownCertCount(count($knownCerts));
        for($x = 0; $x < count($knownCerts); $x++){
          $cert = ""; $pass = "";
          $delimitIdx = strpos($knownCerts[$x], ";");
          if($delimitIdx > 0){
            $cert = substr($knownCerts[$x], 0, $delimitIdx);
            $pass = substr($knownCerts[$x], $delimitIdx+1);
          } else {
            $cert = $knownCerts[$x];
          }

          try {
            $certmgr->doImportFromFile($cert, $pass);
          } catch (exception $e) {
            echo "<h2>Failed to load certificate</h2><p>" . $e->getMessage() . "</p>";
            return;
          }
          $officeverifier->setKnownCertHandle($x, $certmgr->getCertHandle());
        }
      }

      // Trusted certificates
      $certPaths = trim($_REQUEST['trustedCerts']);
      if (strlen($certPaths) > 0) {
        $trustedCerts = explode("\r\n", $certPaths);
        $officeverifier->setTrustedCertCount(count($trustedCerts));
        for($x = 0; $x < count($trustedCerts); $x++){
          $cert = ""; $pass = "";
          $delimitIdx = strpos($trustedCerts[$x], ";");
          if($delimitIdx > 0){
            $cert = substr($trustedCerts[$x], 0, $delimitIdx);
            $pass = substr($trustedCerts[$x], $delimitIdx+1);
          } else {
            $cert = $trustedCerts[$x];
          }

          try {
            $certmgr->doImportFromFile($cert, $pass);
          } catch (exception $e) {
            echo "<h2>Failed to load certificate</h2><p>" . $e->getMessage() . "</p>";
            return;
          }
          $officeverifier->setTrustedCertHandle($x, $certmgr->getCertHandle());
        }
      }

      // Verification
      $officeverifier->doVerify();

      echo "<p>There were " . $officeverifier->getSignatureCount() . " signatures.</p><br />";
      for ($x = 0; $x < $officeverifier->getSignatureCount(); $x++) 
      {
        echo "<h3>Signature #" . ($x + 1) . "</h3><br /><table>";
        
        echo "<tr><td>Signature type:</td><td>" . translateSignatureType($officeverifier->getSignatureSignatureType($x)) . "</td></tr>";
        echo "<tr><td>" . translateDocSig($officeverifier->getSignatureDocumentSigned($x)) . "</td><td></td></tr>";
        echo "<tr><td>" . translateCore($officeverifier->getSignatureCorePropertiesSigned($x)) . "</td><td></td></tr>";
        echo "<tr><td>" . translateOrigSig($officeverifier->getSignatureSignatureOriginSigned($x)) . "</td><td></td></tr>";
        echo "<tr><td>Signature Validation Result:</td><td>" 
                                                . translateSigValidationResult($officeverifier->getSignatureSignatureValidationResult($x))
                                                . "</td></tr>";
        echo "<tr><td>Chain Validation Result:</td><td>" 
                                                . translateChainValidationResult($officeverifier->getSignatureChainValidationResult($x))
                                                . "</td></tr>";
        echo "</table><br />";
      }
    }
    catch (exception $e) {
      echo "<h2>Verification Failure (Details Below)</h2><p>" . $e->getMessage() . "</p>";
    }
  }
?>